cmat/output_generation/report.py

```diff
@@ -0,0 +1,138 @@
+        self.complete_evidence_string_count = 0
+
+        # ClinVar record counters.
+        self.clinvar_total = 0
+        self.clinvar_fatal_no_valid_traits = 0
+        self.clinvar_skip_unsupported_variation = 0
+        self.clinvar_skip_no_functional_consequences = 0
+        self.clinvar_skip_missing_efo_mapping = 0
+        self.clinvar_skip_invalid_evidence_string = 0
+        self.clinvar_done_one_complete_evidence_string = 0
+        self.clinvar_done_multiple_complete_evidence_strings = 0
+        self.clinvar_fatal = 0
+        self.clinvar_skipped = 0
+        self.clinvar_done = 0
+
+        # Total number of trait-to-ontology mappings present in the database.
+        self.total_trait_mappings = 0
+        if trait_mappings:
+            self.total_trait_mappings = sum([len(mappings) for mappings in trait_mappings.values()])
+        # All distinct (trait name, EFO ID) mappings used in the evidence strings.
+        self.used_trait_mappings = set()
+        # All unmapped trait names which prevented evidence string generation and their counts.
+        self.unmapped_trait_names = Counter()
+
+        # Variant-to-consequence mapping counts.
+        self.total_consequence_mappings = 0
+        if consequence_mappings:
+            self.total_consequence_mappings = sum([len(mappings) for mappings in consequence_mappings.values()])
+        self.repeat_expansion_variants = 0
+        self.structural_variants = 0
+
+    def __eq__(self, other):
+        if not isinstance(other, Report):
+            return NotImplemented
+        return self.__dict__ == other.__dict__
+
+    def __add__(self, other):
+        if not isinstance(other, Report):
+            return NotImplemented
+        result = Report()
+        for var_name in vars(self).keys() | vars(other).keys():
+            if var_name == 'total_trait_mappings':
+                result.total_trait_mappings = max(self.total_trait_mappings, other.total_trait_mappings)
+            if var_name == 'total_consequence_mappings':
+                result.total_consequence_mappings = max(self.total_consequence_mappings,
+                                                        other.total_consequence_mappings)
+            if var_name == 'used_trait_mappings':
+                result.used_trait_mappings = self.used_trait_mappings | other.used_trait_mappings
+            else:
+                result.__setattr__(var_name, self.__getattribute__(var_name) + other.__getattribute__(var_name))
+        return result
+
+    def dump_to_file(self, dir_out, filename=COUNTS_FILE_NAME):
+        with open(os.path.join(dir_out, filename), 'w') as f:
+            yaml.safe_dump(vars(self), f)
+
+    def load_from_file(self, filename):
+        with open(filename, 'r') as f:
+            data = yaml.safe_load(f)
+            self.__dict__.update(**data)
+        # yaml loads a dict, convert to counter
+        self.unmapped_trait_names = Counter(self.unmapped_trait_names)
```

Comment:
I was going to say that [pickling](https://docs.python.org/3/library/pickle.html) the objet might be safer but I 'm guessing that you want to use the final output yaml elsewhere?

Reply:
Yes, I was thinking having it in yaml might be more readable and flexible for downstream purposes (e.g. updating the metrics or generating sankey diagrams), but it's probably overkill if those processes end up implemented in Python anyway... I'd probably keep it as it is for now.